그룹 정보 페이지에 데이터 출력하고 댓글 생성, 삭제하기

sanoopark opened the pull request:
### 작업내용
#### 그룹 정보 페이지에 필요한 데이터 출력하기
- `study_group_id`를 통해서 `study_group`을 `SELECT` 했습니다.
- `study_group_id`가 메인 페이지에서 아직 넘어오지 않기 때문에 수동으로 값을 넣었습니다.

#### 기존 멤버라면 `applyBtn`을 비활성화하기
- 유저가 해당 그룹의 기존 멤버인지 확인하기 위해 `study_group_id`를 통해서 `group_member`를 `SELECT` 했습니다.
- `checkIsMember` 함수를 통해 `SELECT`한 `user_id`와 `req.user.id`를 비교합니다.

#### 유저가 `applyBtn`을 클릭하면 필요한 데이터만 전송한 뒤 비활성화하고, 해당 그룹의 멤버로 추가하기 
- 리로드 없이 구현하기 위해서 `submit`을 사용하지 않고 `fetch`를 사용했습니다.
- `applyBtn`을 클릭할 때 `groupId`를 넘기기 위해서 `<button> value`에 값을 넣었습니다.
- `postGroupId` 함수를 통해서 `/group/post-group-id`로 `groupId`를 넘겨줍니다.
- 이 때 `request.payload`의 값을 `parse`하기 위해서 `express.text()` 미들웨어를 사용했습니다.
- `/group/post-group-id`에서는 넘어온 `groupId`를 통해 현재 유저를 그룹 멤버에 추가합니다.

youngToMaturity committed:
feat: mypage 로드시, 로그인 된 user가 가입/생성한 그룹의 데이터를 fetch

각 스터디 클릭시 group-edit에 group_id를 넘길 수 있도록 임시로 실행하고자 하는 과정. (mypage가 완성되면 수정 예정)

youngToMaturity committed:
feat: fetch를 통해 받은 study_gorup_id를 각 group-list__edit-btn의 value로 지정

[임시]
! issue : 각 lead-group-list의 group-list__item의 갯수를 어떻게 지정할 것인가? - user마다 운영하고 있는 스터디 그룹의 갯수가 다를것이기 때문에
! issue: group-list__item에 해당 study_group_id까지는 지정할 수 있겠지만, 해당 div의 내용들을 어떻게 바꿀 것인가? (fetch속의 fetch?)

youngToMaturity committed:
feat : 관리 버튼 클릭시 각 item에 해당하는 study-group-id post

관리 버튼 클릭시 group-list-hidden__input에 저장된 study-group-id가 post된 후 group-edit이 render되도록 함

youngToMaturity committed:
feat: '관리' 버튼 클릭시 post로 group-edit.ejs에 선택된 그룹의 정보 넘기기

db.query를 활용하여 submit된 id와 같은 study_group을 찾고, 해당 그룹에 속해있는 group_member도 result로 저장. 하지만 ejs에 대한 이해도 부족으로 이 값들을 ejs 상에 표시하는것 미완성

42KIM committed:
feat: Load lead-group list by ssr+csr

youngToMaturity committed:
style :  ejs 적용과정 공유를 위한 커밋

youngToMaturity committed:
feat : group-edit 페이지에서 기존의 스터디 그룹 정보를 기본값으로 표시해주기

수정하기 페이지에 들어가는 경우, 해당 스터디의 기존 정보를 page에 보이도록 하였습니다. ( 중분류, 멤버 미구현 )

youngToMaturity committed:
feat: group-edit 페이지 상의 중분류 select를 수정하고자 하는 스터디 그룹의 기존 값으로 지정

route/mypage.js의 /fetch를 /fetch-group으로 수정
js/group-edit.js의 categoryChange(e)에 있던 중분류 list를 전역으로 선언하여 DOMContentLoad시와 change시 모두 활용할 수 있도록 함.
categotyChange(e) 함수내부의 중분류 Element들의 생성과 지정을 setSub()함수로 처리하였고, hidden Input을 통해 확인할 수 있는 render시에 받은 sub과 같은 경우 selected 옵션을 추가.

swprk committed:
feat: Render 'group-info' with data

42KIM committed:
refactor: Change MySQL BETWEEN to LIMIT

42KIM committed:
style: Change variable name

youngToMaturity committed:
feat: 멤버의 수만큼 div를 생성하여 닉네임과 강퇴버튼을 부여

강퇴 버튼의 value에 group-member의 id를 지정했고, 닉네임은 각 회원의 닉네임으로 지정했다. 현재는 임시로 방장도 출력했지만 추후에는 방장을 제외하고 출력하도록 처리할 것

42KIM committed:
feat: 마이페이지 다중쿼리, 서브쿼리 추가

swprk committed:
fix: Handle exception for req.user

ghost commented:
### 추가내용
#### 예외 처리
- 로그인 안 한 상태로 `/group/info/test`로 접근할 때, `req.user.id`가 `undefined`이므로 `app crash` 되어서 예외 처리했습니다.
```javascript
const { id } = req.user ? req.user : '';
```

42KIM committed:
feat: 무한스크롤 기능 구현, 더보기 버튼 삭제

42KIM committed:
feat: Enable create-panel render, link

youngToMaturity committed:
feat : 수정완료 버튼 클릭시 db 수정, 예외처리 두가지 적용

만약 db상에 수정하고자 하는 이름이 있고 && 그 이름이 현재 수정하고자 하는 스터디 그룹의 이름이 아니라면 수정불가.
만약 현재 스터디 인원수보다 수정하고자 하는 스터디 정원이 더 적다면 수정불가.
그 외의 상황은 수정 가능

youngToMaturity committed:
feat: 강퇴 버튼 클릭시 해당 멤버 강퇴 기능 구현

group-edit.ejs - form 위치 수정 (css 재지정 필요)
routes/mypage.js - kickoutForm post 처리, 방장인 경우(추후 방장은 members에 보이지 않도록 처리 요망) 강퇴안됨
js/mypage.js - Kickout시 방생되는 status 2개 추가, 기존의 complete은 edit으로 수정 ?: 각 case 처리 이후에 redirect할 수 있을까?
js/group-edit.js 강퇴 버튼 클릭시 kickout 폼 submit

42KIM committed:
feat: 메인페이지 search-list 구현

42KIM committed:
feat: DB에 생성된 스터디가 없을 때 에러 처리

42KIM committed:
fix: 데이터 없는 경우 분기로 생성

youngToMaturity committed:
Merge remote-tracking branch 'origin/ssr+csr' into group-edit

# Conflicts:
#	public/js/mypage.js
#	routes/mypage.js
#	views/includes/mypage/lead-group-list.ejs

swprk committed:
feat: Add comment section to group-info

swprk committed:
fix: Modify sql query

swprk committed:
fix: Resolve conflict

ghost commented:
### 추가내용
#### 댓글 생성하기
- 최초 SSR 때 모든 댓글을 그려줍니다.
- 등록 버튼을 클릭하면 CSR을 통해 댓글을 그려주고, 해당 `comment` 테이블을 생성합니다.

#### 댓글 삭제하기
- 삭제 버튼을 클릭하면 CSR을 통해 댓글을 지우고, 해당 `comment` 테이블을 삭제합니다.
- `commentUserId`와 `loggedInUserId`를 비교해 해당 댓글을 생성한 유저만 삭제할 수 있습니다.

![example](https://user-images.githubusercontent.com/81365896/127590670-aae5b2fd-b837-484e-a045-54687df4f687.gif)

42KIM committed:
style: Change button tag to input tag, remove classNames

42KIM committed:
feat: Enable mainpage category changing and search-list rendering

42KIM commented:
확인 및 삼국통일(?)  완료!

42KIM reviewed (approved):
확인 및 삼국통일(?) 완료.
기존 
ssr+csr
group-edit
브랜치는 이제 group-info 브랜치로 일원화 하여 운영하는 것으로.

42KIM committed:
Merge branch 'group-info' into mainpage

42KIM committed:
refactor: Change button tags to input in category-filter

42KIM committed:
fix: 관리 버튼 클릭 시 스터디 관리 페이지로 이동 못하는 문제 해결

42KIM committed:
Merge branch 'group-info' into mainpage

42KIM committed:
feat: Add search-list link to group-info page

youngToMaturity commented on routes/group.js:
```diff
@@ -1,19 +1,139 @@
+  );
+});
+
+router.post('/post-group-id', (req, res, next) => {
+  const group_member_id = nanoid();
+  const study_group_id = req.body;
+  const { id: user_id, nickname } = req.user;
+  const sql_member =
+    'INSERT INTO group_member (id, user_id, study_group_id, is_manager, nickname) VALUES (?, ?, ?, ?, ?)';
+  db.query(
+    sql_member,
+    [group_member_id, user_id, study_group_id, 0, nickname],
+    err => {
+      if (err) {
+        next(err);
+      }
+    }
+  );
+});
```
push 받아서 테스트 한 결과 group_member가 새로 생기지 않는데, 상우님 코드에서는 정상 작동 되시나요?
그리고 멤버 추가 과정에 study_group의 current_number와 maximum_number를 비교하여 멤버 추가 가능시에만 추가되는 부분과 추가 이후에 current_number도 +1 시켜주는 과정이 필요할 것 같습니다!

sanoopark replied:
왜 안 되나 했더니 study_group_name 컬럼 추가하면서 쿼리를 변경 안 했네요 확인했습니다!

youngToMaturity committed:
feat : 수정이 정상적으로 처리된 경우에는 mypage로, 그렇지 않은 경우 group-edit 페이지로 redirect

lead-group-list에서 관리 버튼을 클릭하는 경우 form을 submit하여 router.post하던 방식에서 location.href를 통해 group-edit/:id로 이동시킨 후 router.get()을 사용하여 render

youngToMaturity committed:
feat : group-edit의 과정 수정 - 스터디명 중복체크 이후에 수정완료 버튼 활성화, 수정완료 버튼 활성화 이후 스터디명 수정시 다시 비활성화

youngToMaturity committed:
fix : 수정하기 버튼명을 중복체크로 수정

youngToMaturity commented on routes/group.js:
```diff
@@ -1,19 +1,139 @@
+  const group_member_id = nanoid();
+  const study_group_id = req.body;
+  const { id: user_id, nickname } = req.user;
+  const sql_member =
+    'INSERT INTO group_member (id, user_id, study_group_id, is_manager, nickname) VALUES (?, ?, ?, ?, ?)';
+  db.query(
+    sql_member,
+    [group_member_id, user_id, study_group_id, 0, nickname],
+    err => {
+      if (err) {
+        next(err);
+      }
+    }
+  );
+});
```
```suggestion
const sql_study_group = 'SELECT * FROM study_group WHERE id = ?';
  db.query(sql_study_group, [study_group_id], (err, result) => {
    if (err) {
      next(err);
    }
    if (result[0].current_number >= result[0].maximum_number) {
      console.log('정원이 가득 찬 상태. 추가처리 필요');
    } else {
      const sql_member =
        'INSERT INTO group_member (id, user_id, study_group_id, is_manager, nickname, study_group_name, create_date) VALUES (?, ?, ?, ?, ?, ?, now());';
      const sql_update_member =
        'UPDATE study_group SET current_number=? WHERE id=?';
      db.query(
        sql_member + sql_update_member,
        [
          group_member_id,
          user_id,
          study_group_id,
          0,
          nickname,
          result[0].name,
          result[0].current_number + 1,
          study_group_id
        ],
        err => {
          if (err) {
            next(err);
          }
        }
      );
    }
  });
});
```

youngToMaturity replied:
DB상에 정상적으로 처리되는 것 확인하였습니다! 
추가적으로 처리되어야 하는 부분은
* current_number와 maximum_number가 같은 경우, 즉 정원이 가득 차 추가 지원을 할 수 없는 경우에 대한 처리
* group_member 테이블 생성 이후에 DB상의 current_number는 +1된 상태인데 스터디 상세 페이지에서는 +1 이전의 숫자가 표시되는 상태에 대한 처리(CSR로 버튼 클릭시 갱신해주면 될듯?)

sanoopark replied:
아이고.. 감사합니다!
- 스터디 지원했을 때 정원 초과인지 프론트 & 백에서 각각 검증
- 지원 성공했을 때 (정원 초과 아닐 때) DB current_number+1
- 지원 성공했을 때 CSR로 현재 인원 +1
이렇게 3개 작업 완료했고 여기는 클로즈할 거라 따로 PR 올리겠습니다!

swprk committed:
fix: 그룹 지원 과정에서 누락된 부분 추가 - close #91

- 지원 시 정원 초과 검증 추가
- 지원 성공 시 그룹 DB의 현재원 변경
- 지원 성공 시 CSR로 그룹 현재원 변경

42KIM committed:
Merge branch 'group-info' of https://github.com/LM-channel-team-project/wooggooms into group-info

42KIM committed:
Merge branch 'mainpage' into group-info

42KIM commented:
# 8월 16일 Update
mainpage 브랜치를 group-info 브랜치로 merge 하여 push 했습니다.
상우님이 작업 중이신 pdkdf2 브랜치, 영수님이 작업 중이신 Issue/#86 브랜치 외에는 아마 가장 최신화 하여 종합한 브랜치인 것 같습니다.

기존 group-info 페이지에는 없던 메인 페이지의 기능들이 대거 추가되었습니다.
자세한 내용은 mainpage 브랜치를 PR 한 내용을 참고하시면 될 것 같습니다. [👉링크](https://github.com/LM-channel-team-project/wooggooms/pull/82)

@sanoopark @young961027
